src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/nestedfield/NestedFieldProjection.java

```diff
@@ -123,23 +135,64 @@ private void buildInnerHit(List<String> fieldNames, NestedQueryBuilder query) {
      * Why linear search? Because NestedQueryBuilder hides "path" field from any access.
      * Assumption: collected NestedQueryBuilder list should be very small or mostly only one.
      */
-    private NestedQueryBuilder findNestedQueryWithSamePath(List<NestedQueryBuilder> nestedQueries, String path) {
+    private NestedQueryBuilder findNestedQueryWithSamePath(
+            List<NestedQueryBuilder> nestedQueries, String path, List<String> fieldNames) {
+
         return nestedQueries.stream().
-                             filter(query -> isSamePath(path, query)).
-                             findAny().
-                             orElseGet(createEmptyNestedQuery(path));
+                filter(query -> isSamePath(path, query, fieldNames)).
+                findAny().
+                orElseGet(createNestedQuery(path, fieldNames));
+    }
+
+    /** Check against NestedQuery of WHERE condition */
+    private boolean isSamePath(String path, NestedQueryBuilder query, List<String> fieldNames) {
+        boolean isSamePath = nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+        if (isSamePath) {
+           setLastValues(path, fieldNames);
+        }
+        return isSamePath;
     }
 
-    private boolean isSamePath(String path, NestedQueryBuilder query) {
-        return nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+    /**
+     * Since groupFieldNamesByPath() uses TreeMap to maintain reverse alphabetical order. Here we maintain previous
+     * path and fieldName value to compare with current nested query.
+     * */
+    private boolean hasSameParent(String path, QueryBuilder query) {
+        // query could also be termQueryBuilder
+        if (!lastFieldInPath.isEmpty() && (query instanceof NestedQueryBuilder)) {
```

Comment:
I highly recommend to internalize the https://blog.codinghorror.com/flattening-arrow-code/

and embrace ealier exits

if (lastFieldInPath.isEmpty() || ! (query instanceOf NestedQueryBuilder)) {
    return false;
}

main logic

Reply:
Agreed! will apply to the logic.